### PR TITLE
WIP: Data Array

### PR DESCRIPTION
**Pull request by cortner**
CC @jameskermode 

The first problem that I encountered is this: if I only define `get_data`, then based on the key only how do I determine where data is stored? Two options:
1. just search through all three dictionaries (`arrays`, `info`, `transient`)
2. keep a list of all keys that remembers where each of the things are stored.

At the moment, I simply defined `get_data, get_array, get_transient`

### CODE REVIEW

**Comment by cortner** [comment]
I won't merge this quite yet.

**Comment by cortner** [comment]
I tested version 1., the overhead seems to minimal, this might be the best option if we *do* want to combine `info, array, transient` into `data`

**Comment by jameskermode** [comment]
Agreed, I think simply looking through all three dictionaries would be best approach. (Even if we don't do this, then the ``get_data()`` implemented in this PR should be renamed ``get_info()``).

Might also be slightly faster to look directly in ``at.po[:arrays]`` and ``at.po[:info]`` python dictionaries rather than calling ``get_array()`` and ``get_info()`` etc.

**Comment by jameskermode** [comment]
Do we need stubs implementations of some kind for ``AbstractAtoms``?

**Comment by cortner** [comment]
The `abstraction.jl` file basically lists and documents all the functions that *ought* to be implemented; see `@protofun`. I'd be open to implementing a pure Julia `Atoms`, but more as a fallback if `ASE` is not installed.

**Comment by cortner** [comment]
* [ ] Todo before I merge this: `***_momenta` and add a check that the `name` or `id` can never be `positions` or `momenta`.

On this note: do we want to provide `momenta`, `velocities` or both?

**Comment by jameskermode** [comment]
The check is not really needed, as positions and momenta are stored in the same place by ASE, so accessing via ``arrays`` is fine. 

ASE stores momenta and masses and computes velocities if/when requested. I guess we should do the same unless we can think of a reason why not.

**Comment by cortner** [comment]
ok, this is probably ready now.

File: src/ASE.jl [hunk 1]
```diff
@@ -234,7 +236,24 @@ function get_data(at::ASEAtoms, name)
...
-
+# special arrays: momenta, velocities, masses, chemical_symbols
+"get the momenta array"
+momenta(at::ASEAtoms) = at.po[:get_momenta]()
+"set the momenta array"
+set_momenta!(at::ASEAtoms, p::JVecsF) = at.po[:set_momenta](p|>mat)
+"get the velocities array (convert from momenta)"
+velocities(at::ASEAtoms) = at.po[:get_velocities]()
+"convert to momenta, then set the momenta array"
+set_velocities!(at::ASEAtoms, v::JVecsF) = at.po[:get_velocities](v|>mat)
```
Comment (jameskermode) [inline src/ASE.jl]: typo here, should be `set_velocities` on RHS?

**Comment by jameskermode** [comment]
Looks good, thanks, I will test and merge later today if you're happy?

**Comment by cortner** [comment]
that would be great, thank you.
C

File: src/ASE.jl (hunk 1 above)
Comment (cortner) [inline src/ASE.jl]: thank you - fixed.

**Comment by jameskermode** [comment]
There were some missing transpositions going to/from momenta and velocities arrays. Should now be fixed. I'll go ahead and merge now and minor fixes can be done in master.

**Comment by cortner** [comment]
Thank you for checking this so carefully

**Pull request merged**